### PR TITLE
Re-introducing checking of Reason

### PR DESCRIPTION
**Pull request by ImkoMarijnissen**
For a while, the code for checking the reason for a propagation has been unused; likely due to the fact that the reasons which were provided to the method were not `PropositionalConjunction`s anymore.

This PR aims to address this by calculating the reason eagerly (if the assertion levels are set to extreme) and checking these reasons. This already led to an error becoming apparent in the integer multiplication propagator due to an off-by-one error!

### CODE REVIEW

File: pumpkin-lib/src/engine/constraint_satisfaction_solver.rs [hunk 1]
```diff
@@ -1983,65 +1895,66 @@ mod tests {
...
-            CSPSolverStateInternal::Conflict {
-                conflict_info: StoredConflictInfo::Propagation {
-                    reference: _,
-                    literal: _,
-                },
-            }
-        ));
-    }
+    // TODO: test does not work with the new checking of propagations
+    // #[test]
```
Comment (maartenflippo) [inline pumpkin-lib/src/engine/constraint_satisfaction_solver.rs:1899]: We should either change the test or delete it, but not leave it here. I don't see us ever getting back to it if we leave it commented

**Review by maartenflippo — CHANGES_REQUESTED** [review]

File: pumpkin-lib/src/engine/constraint_satisfaction_solver.rs [hunk 2]
```diff
@@ -1422,6 +1431,50 @@ impl ConstraintSatisfactionSolver {
...
+    /// 1. Setting the reason for a propagation should lead to the same propagation when debug
+    ///    propagating from scratch
+    /// 2. Setting the reason for a propagation and the negation of that propagation should lead to
+    ///    failure
+    fn debug_check_propagations(
+        &mut self,
+        num_trail_entries_before: usize,
+        propagator_id: PropagatorId,
+    ) -> bool {
+        {
```
Comment (ImkoMarijnissen) [inline pumpkin-lib/src/engine/constraint_satisfaction_solver.rs:1446]: Unnecessary block?

File: pumpkin-lib/src/engine/debug_helper.rs [hunk 3]
```diff
@@ -149,52 +159,31 @@ impl DebugHelper {
...
-        propagator_id: u32,
+        propagator_id: PropagatorId,
     ) -> bool {
         let reason: PropositionalConjunction = reason
             .iter()
             .cloned()
             .filter(|&predicate| predicate != Predicate::True)
             .collect();
 
-        if reason
```
Comment (ImkoMarijnissen) [inline pumpkin-lib/src/engine/debug_helper.rs:168]: This should never be the case, re-add this check

File: pumpkin-lib/src/engine/constraint_satisfaction_solver.rs [hunk 4]
```diff
@@ -2296,4 +2146,130 @@ mod tests {
             }
         }
     }
+
+    //     struct TestPropagator {
```
Comment (ImkoMarijnissen) [inline pumpkin-lib/src/engine/constraint_satisfaction_solver.rs:2150]: Please Remove

File: pumpkin-lib/src/engine/constraint_satisfaction_solver.rs [hunk 5]
```diff
@@ -1422,6 +1431,44 @@ impl ConstraintSatisfactionSolver {
...
         }
     }
+
+    /// Checks whether the propagations of the propagator since `num_trail_entries_before` are
+    /// reproducible by performing 2 checks:
+    /// 1. Setting the reason for a propagation should lead to the same propagation when debug
+    ///    propagating from scratch
+    /// 2. Setting the reason for a propagation and the negation of that propagation should lead to
+    ///    failure
+    fn debug_check_propagations(
```
Comment (maartenflippo) [inline pumpkin-lib/src/engine/constraint_satisfaction_solver.rs:1441]: This should be in the debug helper I assume?

**Review by maartenflippo — CHANGES_REQUESTED** [review]

**Review by maartenflippo — APPROVED** [review]

**Pull request merged**